dask/array/core.py

```diff
@@ -2009,6 +2019,147 @@ def from_array(x, chunks, name=None, lock=False, asarray=True, fancy=True,
+    """Load array from the zarr storage format
+
+    See https://zarr.readthedocs.io for details about the format.
+
+    Parameters
+    ----------
+    url: str or MutableMapping
+        Location of the data. A URL can include a protocol specifier like s3://
+        for remote data. Can also be any MutableMapping instance, which should
+        be serializable if used in multiple processes.
+    component: str or None
+        If the location is a zarr group rather than an array, this is the
+        subcomponent that should be loaded, something like ``'foo/bar'``.
+    storage_options: dict
+        Any additional parameters for the storage backend (ignored for local
+        paths)
+    chunks: tuple of ints or tuples of ints
+        Passed to ``da.from_array``, allows setting the chunks on
+        initialisation, if the chunking scheme in the on-disc dataset is not
+        optimal for the calculations to follow.
+    kwargs: passed to zarr's open functions.
+    """
+    import zarr
+    storage_options = storage_options or {}
+    if isinstance(url, str):
+        fs, fs_token, path = get_fs_token_paths(
+            url, 'rb', storage_options=storage_options)
+        assert len(path) == 1
+        mapper = get_mapper(fs, path[0])
+    else:
+        mapper = url
+    if component is None:
+        z = zarr.open_array(mapper, mode='r', **kwargs)
+    else:
+        z = zarr.open_group(mapper, mode='r', **kwargs)[component]
+    chunks = chunks if chunks is not None else z.chunks
+    return from_array(z, chunks, name='zarr-%s' % url)
+
+
+def to_zarr(arr, url, component=None, storage_options=None,
+            overwrite_group=False, rechunk=False, compute=True,
+            return_stored=False, **kwargs):
+    """Save array to the zarr storage format
+
+    See https://zarr.readthedocs.io for details about the format.
+
+    Parameters
+    ----------
+    arr: dask.array
+        Data to store
+    url: str or MutableMapping
+        Location of the data. A URL can include a protocol specifier like s3://
+        for remote data. Can also be any MutableMapping instance, which should
+        be serializable if used in multiple processes.
+    component: str or None
+        If the location is a zarr group rather than an array, this is the
+        subcomponent that should be created/over-written.
+    storage_options: dict
+        Any additional parameters for the storage backend (ignored for local
+        paths)
+    overwrite_group: bool
+        If component is not none, this controls whether to create/overwrite the
+        zarr group, or if the group must exits already (but the specific
+        dataset is always overwritten)
+    rechunk: see ``Array.rechunk``
+        Rechunking to be applied to the array before storage, since zarr
+        requires a regular chunks scheme - passed to ``.rechunk()``.
+        If False, no rechunk operation is performed;
+        if the chunks are not regular, an exception is raised.
```

Comment:
I find that this text now accurately describes the situation, and says nothing about *how* to rechunk, only that you might need to do it. The exception is readable, I'd like to think. Rechunking for storage should always be allowed, I think, and any function that passes on to `store` should allow it.